### PR TITLE
realmd: init

### LINKED ISSUE #180043: Realmd missing from SSSD
Hello,

To join a machine to an AD or LDAP domain, SSSD relies on the `realmd` package which is a userspace cli tool to perform the one time join operation (done imperatively).

This is currently missing and consequently makes joining a nixos machine to a domain non-trivial.

Realmd is a freedesktop project (I'm assuming with RedHat blessings)

Package metadata:
- URL: https://www.freedesktop.org/software/realmd/
- [Build Instructions](https://www.freedesktop.org/software/realmd/contribute.html)
- LICENSE: [LGPLv2.1](https://cgit.freedesktop.org/realmd/realmd/tree/COPYING)

Please let me know if any other information is needed

### PR DESCRIPTION
**Pull request by SohamG**
# Description of changes

Fix #180043 

Status Update 24 Nov 2023: [See this comment](https://github.com/NixOS/nixpkgs/pull/181145#issuecomment-1826176185)

## Things done

- Add realmd package.
- Add necessary nixos module.

# Pending changes:
- [ ] Enable polkit and sssd as a part of our enable option.
- [ ] Fix config file paths (perhaps use /run/current-system)
- [ ] More testing

----

- Built on platform(s)
  - [X] x86_64-linux

### CODE REVIEW

**Comment by SohamG** [comment]
I'm sorry for the mess I just had a really bad time with git

**Comment by s1341** [comment]
@SohamG How can I test this? What configuration do I need?

**Comment by s1341** [comment]
Ok. I managed to test with nixpkgs-review... It appears that the dbus service is not brought up.

**Comment by s1341** [comment]
Cool... We also need to change the `realmd-distro.conf` so that its paths make sense for nixos...

**Comment by s1341** [comment]
Yeah. I got `realm discover` working. I need `adcli` and `oddjob`, `oddjob-mkhomedir` too...

**Comment by SohamG** [comment]
> Yeah. I got `realm discover` working. I need `adcli` and `oddjob`, `oddjob-mkhomedir` too...

I believe one derivation per PR is whats expected. Do you think its wise to work those packages out first before pushing on this one?

**Comment by s1341** [comment]
Yeah. One derivation per PR. So let's open separate PRs for those things.

**Comment by s1341** [comment]
I'll try to work on `oddjob` and `oddjob-mkhomedir` tomorrow.

**Comment by SohamG** [comment]
> I'll try to work on `oddjob` and `oddjob-mkhomedir` tomorrow.

This oddjob stuff seems really obscure, [last commit](https://pagure.io/oddjob/commits/master) being 2 years ago.

**Comment by s1341** [comment]
I managed to get the oddjob stuff installed, but can't get it to run properly.... It appears to fail at registering the dbus service.

**Comment by s1341** [comment]
https://github.com/NixOS/nixpkgs/pull/181366

**Comment by SohamG** [comment]
Currently working on the config file which needs nix paths. It looks like its gonna end up being an ugly gawk/sed situation

**Comment by s1341** [comment]
why not use `SubstituteInPlace`?

File: pkgs/os-specific/linux/realmd/default.nix
Comment (s1341) [inline pkgs/os-specific/linux/realmd/default.nix:49]: looks like you're missing an `{` here.

Comment (SohamG) [reply pkgs/os-specific/linux/realmd/default.nix:49]: I removed the { from ${adcli} because adcli isnt merged yet, so it will try to access a variable that doesnt exist

Comment (SohamG) [reply pkgs/os-specific/linux/realmd/default.nix:49]: When adcli is merged, I will uncomment the `if` block and add the `{`

File: pkgs/os-specific/linux/realmd/default.nix
Comment (AndersonTorres) [inline pkgs/os-specific/linux/realmd/default.nix:53]: Boi, such a huge script would be more useful if you did it a separate file and included it.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by SohamG** [comment]
@AndersonTorres Hey, please can you review https://github.com/NixOS/nixpkgs/pull/181366 ? This is actually a dependency for this PR. Once this is merged, this PR should be ready with minimal effort

**Comment by kd2flz** [comment]
@SohamG -
Just wondering where you got to with this - did you hit roadblocks? Just interested...
Thanks! It's good work :)

**Comment by SohamG** [comment]
> @SohamG - Just wondering where you got to with this - did you hit roadblocks? Just interested... Thanks! It's good work :)

Hey, I just got caught up in university work once the semester started. As far as I can remember this is like 90% done , though I dont remember what I was working on last. I will pick this up once I have some free time.

**Comment by kd2flz** [comment]
Thanks, that's good to hear, and good luck with your studies.

________________________________

**Comment by Vinetos** [comment]
Hey, Do you have any test procedure? So I can have a look in the next few days

**Comment by SohamG** [comment]
Hey, from what I recall the only real testing method is to get a VM and to rebuild with my fork of nixpkgs as the source nixpkgs (probably much easier as a flake than standard) and then to attempt to use realmd to join a pre existing ActiveDirectory etc domain. 

I'll try to rebase my fork to the latest to make it a bit easier, and I dont mind handing out write access if anyone wants to chip in.

**Comment by SohamG** [comment]
I have tidy'd up the commits and marked ready for review. Please feel free to test this out in a VM with an existing AD/FreeIPA. As it stands, joining a domain is done **imperatively** with the `realmd join` command. I will try to test this out in the coming days and report back (if I get around to it)

**Comment by brakenium** [comment]
I'm thinking about installing NixOS on my new laptop that is on order and would like to join it to my MS Active Directory domain. Is there anything I can do to help test this? Do you have a Nix config I could use to test this in a VM (QEMU/KVM)? Do I need to download your complete fork? It takes really long to do that. I'm new to Nix so I might not be the best person to help test, but I'm willing to learn

**Comment by SohamG** [comment]
brakenium ***@***.***> writes:

> I'm thinking about installing NixOS on my new laptop that is on order and would like to join it to my
> MS Active Directory domain. Is there anything I can do to help test this? Do you have a Nix config I
> could use to test this in a VM (QEMU/KVM)? Do I need to download your complete fork? It takes really
> long to do that. I'm new to Nix so I might not be the best person to help test, but I'm willing to
> learn

Well, you dont need to download the *entire* fork, something like
```
git clone https://github.com/sohamg/nixpkgs --branch pkg/realmd --depth 2
```

Should finish much faster and only download the necessary bits. For
testing, however, the process isnt quite straightforward: You need to
tell nixos-rebuild to use the previously cloned nixpkgs as the nixpkgs
for the system. I believe this is done with the -A option (dont quote me
on that, flake user here).

That being said, I wont recommend testing on a production/in-use
machine. I am testing in a VM right now, will have more to share
soon. If you want, you can help by testing in a **VM** on your new laptop!

**Comment by SohamG** [comment]
Testing notes (Thanksgiving edition):
- I could swear the realmd package would build properly, but I guess not? (its the same hashes!) Added one line fix to preConfigure to get it to build.

- Discovery: Realmd throws a fit if you dont have polkit installed which means anything that doesnt have a GUI. [No ones fixed this yet.](https://bugs.freedesktop.org/show_bug.cgi?id=90683). Need to make the nixos module enable polkit as well.

- The realmd config has broken paths (apparently?) due to double substitution of some kind (from normal linux paths to nix store paths). Will need to fix this.

Hopefully I can work more on this tomorrow. (I will gladly accept help!)

**Comment by brakenium** [comment]
Currently I'm doing some testing on my desktop in a VM. My laptop won't be here until January. I am trying to get a suitable NixOS config setup before it arrives and would like it to be domain joined. Later I remembered I can just curl the repo's zip file. I should be able to test it out if you can get me an example config (assuming you're testing with AD as well)

**Comment by SohamG** [comment]
brakenium ***@***.***> writes:

> Later I remembered I can just curl the repo's zip file. I should be able to test it out if you can
> get me an example config (assuming you're testing with AD as well)

You only need a couple lines in the config:

services.realmd.enable = true;
services.polkit.enable = true;
services.sssd.enable = true;

And then after rebuilding you can join the domain with the usual realm
commands

realm join domain.tld -U adminuser

Currently the join fails cause I need to fix the config file being
generated. Thanks for the willingness to help. I should have it fixed
soon and ready for actual testing.

**Comment by brakenium** [comment]
Sounds good. Please let me know when this is ready to test

File: pkgs/os-specific/linux/realmd/default.nix
Comment (anthonyroussel) [inline pkgs/os-specific/linux/realmd/default.nix:30]:
```suggestion
    hash = "sha256-bpX+0/T6E9ml98rcVQGrVXMHRYHWDPanCk7MaZAS2LA=";
```

sha256 has been deprecated in favor of hash

File: nixos/modules/services/admin/realmd.nix
Comment (anthonyroussel) [inline nixos/modules/services/admin/realmd.nix:5]:
```suggestion
let
  inherit (lib) mkIf mkOption types;
```

Using `with lib;` is not recommended: https://nix.dev/guides/best-practices.html#with-scopes

I recommend you to use this notation instead :)

File: pkgs/os-specific/linux/realmd/default.nix
Comment (anthonyroussel) [inline pkgs/os-specific/linux/realmd/default.nix:30]:
```suggestion
  version = "0.17.1";

  srcs = fetchFromGitLab {
    domain = "gitlab.freedesktop.org";
    owner = pname;
    repo = pname;
    rev = version;
    hash = "sha256-lmNlrXOOUSDk/8H/ge0IRA64bnau9nYUIkW6OyVxbBg=";
```

Looks like the latest released version is 0.17.1

File: pkgs/os-specific/linux/realmd/default.nix
Comment (anthonyroussel) [inline pkgs/os-specific/linux/realmd/default.nix:80]:
```suggestion

  meta = {
    homepage = "https://www.freedesktop.org/software/realmd";
    changelog = "https://gitlab.freedesktop.org/realmd/realmd/-/blob/${version}/NEWS";
    description = "DBus service for configuring kerberos and other online identities.";
    license = lib.licenses.lgpl21Only;
    mainProgram = "realm";
    maintainers = with lib.maintainers; [ SohamG ];
    platforms = lib.platforms.linux;
  };
}
```

File: nixos/modules/services/admin/realmd.nix
Comment (anthonyroussel) [inline nixos/modules/services/admin/realmd.nix:24]: I recommend you to use [RFC0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md) to have structural settings instead of stringly config text:

```diff
   pkg = pkgs.realmd;
   cfg = config.services.realmd;
+
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "realmd.conf" cfg.settings;
 in
 {
   options.services.realmd = {
@@ -18,8 +24,14 @@ in
     };
 
     configText = mkOption {
-      type = types.lines;
-      default = "";
+      type = lib.types.submodule { freeformType = settingsFormat.type; };
+      default = { };
+      example = {
+        users = {
+          default-home = "/home/%U";
+          default-shell = "${pkgs.bash}/bin/bash";
+        };
+      };
       description = "The verbatim contents of config file /etc/realmd.conf";
     };
   };
@@ -41,9 +53,6 @@ in
       };
     };
 
-    environment.etc."realmd.conf" = {
-      target = "realmd.conf";
-      source = pkgs.writeText "realmd.conf" "${cfg.configText}";
-    };
+    environment.etc."realmd.conf".source = configFile;
   };
 }
```

What do you think?

File: nixos/modules/services/admin/realmd.nix
Comment (anthonyroussel) [inline nixos/modules/services/admin/realmd.nix:6]: `pkg` is unused here.

If you want to dynamically pass the realmd package into the realmd NixOS module, you can add this instead:

```nix
package = lib.mkPackageOptionMD pkgs "realmd" {};
```

Then use it later in the ExecStart:

```nix
serviceConfig = {
  Type = "dbus";
  BusName = "org.freedesktop.realmd";
  ExecStart = "${cfg.package}/libexec/realmd";
};
```

What do you think?

**Comment by anthonyroussel** [comment]
Unfortunately the `realm join` command fails with `/usr/bin/authselect` tool not found (which is a Redhat-only tool).


```console
machine # [   13.565952] realmd[769]:  * /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash -c /usr/bin/authselect select sssd with-mkhomedir --force && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl enable oddjobd.service && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl start oddjobd.service
machine # [   13.573824] realmd[769]: process started: 934
machine # [   13.580805] realmd[769]: /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash: line 1: /usr/bin/authselect: No such file or directory
```

**Comment by SohamG** [comment]
> Unfortunately the `realm join` command fails with `/usr/bin/authselect` tool not found (which is a Redhat-only tool).
> 
> ```
> machine # [   13.565952] realmd[769]:  * /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash -c /usr/bin/authselect select sssd with-mkhomedir --force && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl enable oddjobd.service && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl start oddjobd.service
> machine # [   13.573824] realmd[769]: process started: 934
> machine # [   13.580805] realmd[769]: /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash: line 1: /usr/bin/authselect: No such file or directory
> ```
> 
> I think we need to adapt the realmd-distro.conf configuration file for NixOS.

Yes, working on this exactly right now. Will onboard the other suggestions too (including the INI config one)

**Comment by kd2flz** [comment]
> > Unfortunately the `realm join` command fails with `/usr/bin/authselect` tool not found (which is a Redhat-only tool).
> > ```
> > machine # [   13.565952] realmd[769]:  * /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash -c /usr/bin/authselect select sssd with-mkhomedir --force && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl enable oddjobd.service && /nix/store/vapnrxrw3b21c7ji61bmnbzl2cj1vl96-systemd-254.3/bin/systemctl start oddjobd.service
> > machine # [   13.573824] realmd[769]: process started: 934
> > machine # [   13.580805] realmd[769]: /nix/store/lf0wpjrj8yx4gsmw2s3xfl58ixmqk8qa-bash-5.2-p15/bin/bash: line 1: /usr/bin/authselect: No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think we need to adapt the realmd-distro.conf configuration file for NixOS.
> 
> Yes, working on this exactly right now. Will onboard the other suggestions too (including the INI config one)

Just curious what your thoughts were for going forward with this. I think I prefer the INI config method, as it would be more reproducible.

File: nixos/modules/services/admin/realmd.nix
Comment (nu-nu-ko) [reply nixos/modules/services/admin/realmd.nix:5]: why was this marked resolved, the suggestion was not included and there's no reasoning against it.

**Pull request closed without merging**
